keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/SumVertex.java

```diff
@@ -2,42 +2,110 @@
 
 public class SumVertex extends DoubleUnaryOpVertex {
 
+    private final int[] overDimensions;
+
     /**
-     * Performs a sum across each value stored in a vertex
+     * Performs a sum across specified dimensions
+     *
+     * @param inputVertex    the vertex to have its values summed
+     * @param overDimensions dimensions to sum over
+     */
+    public SumVertex(DoubleVertex inputVertex, int[] overDimensions) {
+        super(getSummationResultShape(inputVertex.getShape(), overDimensions), inputVertex);
+        this.overDimensions = overDimensions;
+    }
+
+    /**
+     * Performs a sum across all dimensions
      *
      * @param inputVertex the vertex to have its values summed
      */
     public SumVertex(DoubleVertex inputVertex) {
-        super(Tensor.SCALAR_SHAPE, inputVertex);
+        this(inputVertex, TensorShape.dimensionRange(0, inputVertex.getShape().length));
+    }
+
+    /**
+     * This is here due to strange behavior in tensor summing over dimensions where
+     * dimensions are not dropped if the rank is 2 or less.
+     */
+    private static int[] getSummationResultShape(int[] inputShape, int[] sumOverDimensions) {
+        List<Integer> inputShapeList = new ArrayList<>(Ints.asList(inputShape));
+
+        for (int dim : sumOverDimensions) {
+            inputShapeList.set(dim, 0);
+        }
+
+        for (int i = inputShapeList.size() - 1; i >= 0; i--) {
+            if (inputShapeList.get(i) == 0) {
+                if (inputShapeList.size() > 2) {
+                    inputShapeList.remove(i);
+                } else {
+                    inputShapeList.set(i, 1);
+                }
+            }
+        }
+
+        return Ints.toArray(inputShapeList);
     }
 
     @Override
     protected DoubleTensor op(DoubleTensor value) {
-        return DoubleTensor.scalar(value.sum());
+        return value.sum(overDimensions);
     }
 
     @Override
     protected DualNumber dualOp(DualNumber dualNumber) {
-        return dualNumber.sum();
+        return dualNumber.sum(overDimensions);
     }
 
     @Override
     public Map<Vertex, PartialDerivatives> reverseModeAutoDifferentiation(PartialDerivatives derivativeOfOutputsWithRespectToSelf) {
 
-        PartialDerivatives derivativesWrtInput = derivativeOfOutputsWithRespectToSelf
-            .multiplyAlongWrtDimensions(DoubleTensor.ones(inputVertex.getShape()), this.getShape());
+        int[] wrtShape = summedOverShapeWithoutRankLoss(inputVertex.getShape(), overDimensions);
+
+        PartialDerivatives reshapedDiffWrtSelf = new PartialDerivatives(new HashMap<>());
+        for (Map.Entry<VertexId, DoubleTensor> partialDerivative : derivativeOfOutputsWithRespectToSelf.asMap().entrySet()) {
+            DoubleTensor partial = partialDerivative.getValue();
+
+            int[] newPartialShape = TensorShape.concat(
+                TensorShape.selectDimensions(0, partial.getRank() - getShape().length, partial.getShape()),
```

Comment:
can we pull this apart a bit to make it clear what all the bits mean?

Reply:
👍